src/main/java/com/google/devtools/build/lib/actions/PrecomputedPathMapper.java

```diff
@@ -0,0 +1,358 @@
+   * <p>
+   * Compared to full config stripping, where an input file under the path
+   * `bazel-out/k8-fastbuild/bin/pkg/name` would be staged as `bazel-out/bin/pkg/name`, the
+   * per-action scheme has two advantages:
+   * 1. Actions which have artifacts among their inputs that agree in the root-relative path
+   *    portion, but differ in the config part, can still benefit from improved caching.
+   * 2. The current scheme preserves the property of config-based roots that roots combined with the
+   *    root-relative paths of artifacts uniquely identify a particular file content in the entire
+   *    build. As a result, the path mapping can be "materialized" in the local exec root build-wide
+   *    without creating ambiguities if needed, allowing for e.g. debuggers to resolve the paths of
+   *    both source and generated input files.
+   *
+   * <p>
+   * Note: The paths at which inputs are stages with per-action path mapping are only valid in the
+   * context of a single action. Per-action path mapping thus shouldn't be used for actions that
+   * e.g. write full paths to a manifest that is meant to be consumed by another action.
+   */
+  public static PathMapper createPerAction(
+      ArtifactExpander artifactExpander,
+      MetadataHandler metadataHandler,
+      NestedSet<Artifact> inputs,
+      Artifact primaryOutput) throws CommandLineExpansionException {
+    List<Artifact> expandDerivedArtifacts = expandDerivedArtifacts(artifactExpander, inputs);
+    if (expandDerivedArtifacts == null) {
+      return NOOP;
+    }
+    Fingerprint fp = new Fingerprint();
+    // A digest of the set of tuples (root-relative path, file digest) for all generated input
+    // artifacts. This digest serves as the base for the replacement of the configuration-specific
+    // part of the output root (e.g. `k8-fastbuild-ST-ABCDEF`).
+    byte[] virtualConfigDigest = new byte[DigestUtils.ESTIMATED_SIZE];
+    // For every root-relative path among input artifacts, collect all generated artifacts with this
+    // path together with their digests.
+    Map<String, ArrayList<Pair<DerivedArtifact, byte[]>>> rootRelativePathCollisions = new HashMap<>();
+    boolean hasRootRelativePathCollision = false;
+    for (Artifact input : expandDerivedArtifacts) {
+      DerivedArtifact derivedArtifact = (DerivedArtifact) input;
+      String rootRelativePath = derivedArtifact.getRootRelativePathString();
+      fp.addString(rootRelativePath);
+      FileArtifactValue md;
+      try {
+        md = metadataHandler.getMetadata(input);
+      } catch (IOException e) {
+        throw new CommandLineExpansionException(
+            String.format("Failed to get metadata for %s", input), e);
+      }
+      Preconditions.checkNotNull(md);
+      byte[] digest = md.getDigest();
+      if (digest == null) {
+        return NOOP;
+      }
+      fp.addBytes(digest);
+      virtualConfigDigest = DigestUtils.xor(virtualConfigDigest, fp.digestAndReset());
```

Comment:
Note: This suffers from the issue explained in https://github.com/bazelbuild/bazel/issues/15660. It can easily be fixed by collecting all pairs in a set first and then hashing it.